petastorm/pytorch.py

```diff
@@ -100,11 +102,42 @@ def decimal_friendly_collate(batch):
 loader."
 
 
+class BackgroundIterator(threading.Thread):
+    """Prefetch iterator results. A thread iterates the original iterator and
+    populates a queue. Iterating over this background iterator just consumes the underlying
+    queue until no other result is available."""
+    def __init__(self, iterator, queue_size=1000):
+        threading.Thread.__init__(self)
+        self.name = "background_iterator"
+        self.queue = Queue(queue_size)
+        self.iterator = iterator
+        self.stop = threading.Event()
+        self.start()
+
+    def run(self):
+        while not self.stop.isSet():
+            for item in self.iterator:
+                self.queue.put(item)
```

Comment:
I don't think we can use blocking puts and gets and end up with a solution that is robust to deadlocks. Let's see if this works:
- Producer ended up filling the queue and waits on a blocking put.
- Consumer fails. Calls `iterator.stop.set()`, however since .put is blocking within an iterator queue, the event is never checked and the thread is not shut down.

Another scenario: 
- The queue is empty, hence consumer waits on a blocking `.get`. 
- However, producer raises an exception. The thread dies and the consumer is stuck forever on a .get.

I think a robust implementation for a `BackgroundIterator` could get pretty tricky. All these edge cases need to be carefully tested as these kind of failures would be hard to catch in production.